File: dart-wrapper/lib/src/commonMain/kotlin/com/icure/cardinal/sdk/dart/auth/Credentials.kt
package com.icure.cardinal.sdk.dart.auth

import com.icure.cardinal.sdk.auth.ThirdPartyProvider
import kotlinx.serialization.Serializable

@Serializable
sealed interface Credentials {

	fun toKt(): com.icure.cardinal.sdk.auth.Credentials

	@Serializable
	data class UsernamePassword(val username: String, val password: String) : Credentials {
		override fun toKt() = com.icure.cardinal.sdk.auth.UsernamePassword(username, password)
	}

	data class UsernameLongToken(val username: String, val token: String) : Credentials {
		override fun toKt() = com.icure.cardinal.sdk.auth.UsernameLongToken(username, token)
	}

	data class ThirdPartyAuthentication(val token: String, val provider: ThirdPartyProvider) : Credentials {
		override fun toKt() = com.icure.cardinal.sdk.auth.ThirdPartyAuthentication(token, provider)
	}

}

File: dart-wrapper/cardinal_sdk/lib/auth/credentials.dart
class Credentials {

  final Map<String, dynamic> _json;

  Credentials._internal(this._json);

  /// Instantiates the credentials to log in using a username and a password.
  /// The [username] can be:
  /// - [User.id]
  /// - [Group.id]:[User.id]
  /// - [User.login]
  /// - [User.email]
  /// The [password] is the user password.
  factory Credentials.UsernamePassword(String username, String password) {
    return Credentials._internal({
      "username": username,
      "password": password,
      "kotlinType": "com.icure.cardinal.sdk.dart.auth.Credentials.UsernamePassword"
    });
  }

  /// Instantiates the credentials to log in using a username and a long-lived token.
  /// The [username] can be:
  /// - [User.id]
  /// - [Group.id]:[User.id]
  /// - [User.login]
  /// - [User.email]
  /// The [token] is a long-lived authentication token generated through the SDK
  /// or through the Cockpit.
  factory Credentials.UsernameLongToken(String username, String password) {
    return Credentials._internal({
      "username": username,
      "password": password,
      "kotlinType": "com.icure.cardinal.sdk.dart.auth.Credentials.UsernameLongToken"
    });
  }

  /// Instantiates the credentials to log in using a third-party authentication provider.
  /// The [provider] is the provider using to log in and [token] is the token
  /// received upon completing the third-party authentication flow.
  factory Credentials.ThirdPartyAuthentication(String token, ThirdPartyProvider provider) {
    return Credentials._internal({
      "provider": ThirdPartyProvider.encode(provider),
      "token": token,
      "kotlinType": "com.icure.cardinal.sdk.dart.auth.Credentials.UsernamePassword"
    });
  }

  static Map<String, dynamic> encode(Credentials value) {
    return Map.unmodifiable(value._json);
  }
}

enum ThirdPartyProvider {
  google;

  static String encode(ThirdPartyProvider value) {
    switch(value) {
      case ThirdPartyProvider.google:
        return "GOOGLE";
    }
  }
}